README.md
# Fourth down calculator

This is the repository for the [fourth down calculator](https://rbsdm.com/stats/fourth_calculator) introduced in [this piece on The Athletic](https://theathletic.com/2144214/2020/10/28/nfl-fourth-down-decisions-the-math-behind-the-leagues-new-aggressiveness/). Here are the main files of interest:

* [Model for yards gained on fourth down go decision](https://github.com/guga31bb/fourth_calculator/blob/main/R/_go_for_it_model.R)
* [Modeling punt and field goal outcomes](https://github.com/guga31bb/fourth_calculator/blob/main/R/punts.R)
* [Code for generating the figures and tables in The Athletic piece](https://github.com/guga31bb/fourth_calculator/blob/main/R/_the_athletic_post.R)
* [The logic of the Shiny app](https://github.com/guga31bb/fourth_calculator/blob/main/app.R) which is located [here](https://rbsdm.com/stats/fourth_calculator/)
* [The functions that do the win probability calculations](https://github.com/guga31bb/fourth_calculator/blob/main/R/helpers.R)

The code that powers the Twitter fourth down bot [is in this folder here](https://github.com/guga31bb/fourth_calculator/tree/main/bot).

To get starting using the code, I would recommend **cloning this repository** (which gets the dataframes with 4th down calculations already applied) and then going through [my file analyzing the  data](https://github.com/guga31bb/fourth_calculator/blob/main/R/season_numbers.R).

## Features

* The **go for it** model gives probabilities for possibilities of yards gained and includes the possibility of earning a first down via defensive penalty
* The **punt** model includes the possibility for getting blocked, returned for a touchdown, or fumbled on the return
* The **field goal** model is a simple model of field goal % by distance and roof type

## Current limitations

There are some edge cases that are not accounted for. These should only make a marginal difference to the recommendations as they are largely edge cases (e.g. the possibility for a field goal to be blocked and returned).

* The **go for it** model does not allow for the possibility of a turnover return. However, long returns are extremely rare: For example, in 2018 and 2019 there were only four defensive touchdowns on plays where teams went for fourth downs out of 1,236 plays, and all of these happened when the game was well in hand for the other team.
* The **punt** model doesn’t account for the punter or returner, ignores penalties on returns and ignores the potential for blocked punts to be returned for touchdowns
* The **field goal** model doesn’t account for who the kicker is, what the weather is (only relevant for outdoor games), or the possibility of a kick being blocked and returned for a touchdown

## Example usage 1: user input

Here is the code that can look up one play. This is the controversial field goal attempt that the Packers attempted at the end of the 2020 NFC Championship Game.

``` r
source('R/helpers.R')

# get the situation from user input
df <- 
    tibble::tibble(
      
      # reg or post
      "type" = "post",
      
      "qtr" = 4,
      
      # since user input is mins and seconds
      "time" = 60 * as.integer(2) + as.integer(9),
      
      'posteam' = as.character("GB"),
      'away_team' = as.character("TB"),
      'home_team' = as.character("GB"),
      'yardline_100' = as.integer(8),
      'ydstogo' = as.integer(8),
      'posteam_timeouts_remaining' = as.integer(3),
      'defteam_timeouts_remaining' = as.integer(3),
      
      # did home team receive opening kickoff?
      'home_opening_kickoff' = as.integer(0),
      
      'score_differential' = as.integer(-8),
      
      # user input for additional runoff in seconds after successful 4th down conversion
      'runoff' = as.integer(0),
      
      # season that the game happened in
      'yr' = as.integer(2020)
    ) %>%
      # helper function to look up point spread and roof type for given game
      prepare_df(games)

# main function in R/helpers.R
# punt_df is automatically loaded by the source line above and contains the distribution of punts

make_table_data(df, punt_df)

# A tibble: 3 x 5
  choice             choice_prob success_prob fail_wp success_wp
  <chr>                    <dbl>        <dbl>   <dbl>      <dbl>
1 Go for it                12.7          32.7    3.53      31.5 
2 Field goal attempt        8.90         97.5    2.99       9.05
3 Punt                     NA            NA     NA         NA 
```

## Example usage 2 : from an nflfastR play

Typing in all the situation variables is a pain. Here's how to get the table from a play that already exists in nflfastR data:

``` r
# helper function that prepares data and most importantly creates indicator for 2nd half kickoff team
source("R/season_numbers_functions.R")
source('R/helpers.R')

# get the play (I cheated and already looked up play ID)
pbp <- readRDS(url("https://raw.githubusercontent.com/guga31bb/nflfastR-data/master/data/play_by_play_2020.rds")) %>%
  prepare_data() %>%
  filter(week == 20, posteam == "GB", play_id == 3728) %>%
  prepare_df(games) %>%
  # this is an nflfastR bug where existing wp messes with the 4th down functions
  select(-vegas_wp)

make_table_data(pbp, punt_df)

# A tibble: 3 x 5
  choice             choice_prob success_prob fail_wp success_wp
  <chr>                    <dbl>        <dbl>   <dbl>      <dbl>
1 Go for it                12.7          32.7    3.53      31.5 
2 Field goal attempt        8.90         97.5    2.99       9.05
3 Punt                     NA            NA     NA         NA   
```
## Example usage 3 : make the table

Here's how to get the actual table on the shiny app and tweeted out by the bot.
```
make_table(make_table_data(pbp, punt_df), pbp)
 
```
![https://pbs.twimg.com/media/Ess2ZxrXAAUgRtZ?format=png&name=small](https://pbs.twimg.com/media/Ess2ZxrXAAUgRtZ?format=png&name=small)

CHANGELOG.md
## 2021-01-28

* Fixes with touchdowns. Instead of granting 7 points, assumes teams choose best option
between PAT or 2pt and give pre-conversion attempt WP accordingly

## 2021-01-23

* Fix for punt WP at the end of first half being too high

## 2020-11-30

* Add logic for TD/FG decision on the last play of the first half

## 2020-11-28 

* Add possibility for muffed punt

## 2020-11-02

* Reduce field goal chances on very long field goals

app.R
source('R/helpers.R')
library(shiny)
library(shinythemes)

# get teams and logos
teams <- nflfastR::teams_colors_logos %>%
  filter(!team_abbr %in% c("LAR", "SD", "STL", "OAK"))

# team abbreviations for dropdown menus
ids <- teams %>%
  pull(team_abbr)
  
#####################################################################################
######## Define UI for viewer app ###########################################
#####################################################################################

ui <- function(request) {
  
  fluidPage(
    
    theme = shinytheme("cerulean"),
    # theme = shinytheme("sandstone"),
    # shinythemes::themeSelector(),
    
    # App title ----
    titlePanel("Fourth Down Decision Calculator"),
    
    # Main panel for displaying outputs ----
    mainPanel(
      
      navbarPage("rbsdm.com/stats",
                 tabPanel("Result",

                          fluidRow(
                            column(12, align = "center",
                                   bookmarkButton()
                            )),
                          fluidRow(
                            column(4, align = "center",
                                   
                                   tags$h3("Game"),

                                   selectInput(
                                     inputId =  "season", 
                                     label = "Season:", 
                                     choices = 2014:2020,
                                     selected = 2020
                                   ),
                                   
                                   radioButtons("type",
                                                label = "Type:",
                                                choices = list("Regular" = "reg","Postseason" = "post"),
                                                selected = "reg"),
                                   
                                   selectInput("posteam",
                                               "Offense:",
                                               c(sort(unique(as.character(ids)))), selected = "TB"),
                                   
                                   selectInput("away",
                                               "Away team:",
                                               c(sort(unique(as.character(ids)))), selected = "TB"),
                                   
                                   selectInput("home",
                                               "Home team:",
                                               c(sort(unique(as.character(ids)))), selected = "CHI"),
                                   
                                   radioButtons("home_ko",
                                                label = "Did home team get opening kickoff?:",
                                                choices = list("0" = 0, "1" = 1),
                                                inline = T,
                                                selected = "0")
                                   
                                   
                                   ),
                            column(4, align = "center",
                                   
                                   tags$h3("Time"),
                                   radioButtons("qtr",
                                                label = "Quarter:",
                                                choices = list("1" = 1, "2" = 2, "3" = 3, "4" = 4),
                                                inline = T,
                                                selected = "4"),
                                   
                                   sliderInput(
                                     inputId =  "mins", 
                                     label = "Minutes left in quarter:", 
                                     min = 0, max = 15,
                                     value = 4
                                   ),
                                   
                                   sliderInput(
                                     inputId =  "secs", 
                                     label = "Seconds left in quarter:", 
                                     min = 0, max = 59,
                                     value = 52
                                   ),

                                   radioButtons("posteam_to",
                                                label = "Offense timeouts:",
                                                choices = list("0" = 0, "1" = 1, "2" = 2, "3" = 3),
                                                inline = T,
                                                selected = "3"),
                                   
                                   radioButtons("defteam_to",
                                                label = "Defense timeouts:",
                                                choices = list("0" = 0, "1" = 1, "2" = 2, "3" = 3),
                                                inline = T,
                                                selected = "3"),
                                   
                                   
                                   
                                   
                                   
                            ),
                            
                            column(4, align = "center",
                                   
                                   tags$h3("Situation"),

                                   sliderInput(
                                     inputId =  "ydstogo", 
                                     label = "Yards to go:", 
                                     min = 1, max = 30,
                                     value = 1
                                   ),
                                   
                                   sliderInput(
                                     inputId =  "yardline", 
                                     label = "Yards from opponent EZ:", 
                                     min = 1, max = 90,
                                     value = 7
                                   ),
                                   
                                   sliderInput(
                                     inputId =  "score_diff", 
                                     label = "Score differential:", 
                                     min = -28, max = 28,
                                     value = -1
                                   ),
                                   
                                   sliderInput(
                                     inputId =  "runoff", 
                                     label = "Advanced: additional seconds to run off clock after successful conversion", 
                                     min = 0, max = 40,
                                     value = 0
                                   )
                                   
                                   
                                   )

                          ),
                          
                          fluidRow(
                            column(12, align = "center",
                                   actionButton("update", "Update", width = '50%')
                            )),
                          
                          ############ start panel
                          
                          tabsetPanel(type = "tabs", id = "inTabset",
                                      tabPanel("4th down",
                                               fluidRow(
                                                 column(12, align = "center",
                                                        tags$br(),
                                                        tags$br(),
                                                        
                                                        htmlOutput("picture"),
                                                        htmlOutput("some_text")
                                                        
                                                 )
                                               ),
                                               
                                               fluidRow(
                                                 
                                                 column(12, align="center", 
                                                        gt_output(outputId = "view1") %>% 
                                                          shinycssloaders::withSpinner(type = 6,# types see https://projects.lukehaas.me/css-loaders/ 
                                                                                       color = "#414141",
                                                                                       color.background = "#FFFFFF")
                                                 )
                                               ),
                                               
                                               tags$br(),
                                               
                                               tags$p("Notes: these are ESTIMATES; please use accordingly. On 4th and 1, the model cannot know whether it's a long 1 or a short 1. Shorter distance would favor going. Do not use this in overtime. Use with EXTREME CAUTION in the final minute of a game as the model is not good with end of game clock scenarios.")
                                               
                                               
                                               ### end panel
                                               
                                               
                                               
                                               
                                      ),
                                      tabPanel("2 point conversions", value = "two_pt",
                                               
                                               fluidRow(
                                                 column(12, align = "center",
                                                        tags$br(),
                                                        tags$br(),
                                                        
                                                        htmlOutput("picture2"),
                                                        htmlOutput("some_text2")
                                                        
                                                 )
                                               ),
                                               fluidRow(
                                                 
                                                 column(12, align="center", 
                                                        gt_output(outputId = "view2") %>% 
                                                          shinycssloaders::withSpinner(type = 6,# types see https://projects.lukehaas.me/css-loaders/ 
                                                                                       color = "#414141",
                                                                                       color.background = "#FFFFFF")
                                                 )
                                               )
                                               
                                               )

                                      )
                          
                                   
                          ),
                 tabPanel("About",
                          p("Notes: Data from @nflfastR. A thank you to Lee Sharpe (@LeeSharpeNFL) for hosting an updating source for game results and to Thomas Mock (@thomas_mock) for code to make the table look better. Website by Ben Baldwin (@benbbaldwin). Model writeup coming at some point...")
                          
                          
                                      )

 
                 
      )
      
      
      
      
    )
    
    
  )
  
}

#####################################################################################
######## Define server app ###########################################
#####################################################################################
server <- function(session, input, output) {
  
  # Return the requested dataset ----
  # Note that we use eventReactive() here, which depends on
  # input$update (the action button), so that the output is only
  # updated when the user clicks the button

  # uncomment only for testing
  # input <- NULL
  # input$qtr <- 4
  # input$mins <- 4
  # input$secs <- 52
  # input$posteam <- "TB"
  # input$away <- "TB"
  # input$home <- "CHI"
  # input$yardline <- 7
  # input$ydstogo <- 1
  # input$posteam_to <- 3
  # input$defteam_to <- 3
  # input$home_ko <- 0
  # input$score_diff <- -1
  # input$type <- "reg"
  # input$runoff <- 0
  # input$season <- 2020
  #

  # get the situation from user input
  fullInput <- eventReactive(
    input$update,
    {
 
      tibble::tibble(
        "type" = as.character(input$type),
        "qtr" = as.integer(input$qtr),
        "time" = 60 * as.integer(input$mins) + as.integer(input$secs),
        'posteam' = as.character(input$posteam),
        'away_team' = as.character(input$away),
        'home_team' = as.character(input$home),
        'yardline_100' = as.integer(input$yardline),
        'ydstogo' = as.integer(input$ydstogo),
        'posteam_timeouts_remaining' = as.integer(input$posteam_to),
        'defteam_timeouts_remaining' = as.integer(input$defteam_to),
        'home_opening_kickoff' = as.integer(input$home_ko),
        'score_differential' = as.integer(input$score_diff),
        'runoff' = as.integer(input$runoff),
        'yr' = as.integer(input$season)
      ) %>%
        prepare_df(games)
      
    } , ignoreNULL = FALSE
  )
  
  # get the data that goes in the table
  tableData <- eventReactive(
    input$update,
    {
      
      make_table_data(fullInput(), punt_df)
      
    } , ignoreNULL = FALSE
  )
  
  # make the table
  output$view1 <- render_gt(
      expr = make_table(tableData(), fullInput())
  )
  
  # get the data that goes in the 2pt table
  tableData2 <- eventReactive(
    input$update,
    {
      
      make_2pt_table_data(fullInput())
      
    } , ignoreNULL = FALSE
  )
  
  # make the 2pt table
  output$view2 <- render_gt(
    expr = make_table_2pt(tableData2(), fullInput())
  )

  # for the team logo above decision
  output$picture <-
    renderText({
      c(
        '<img width="150" src="',
        glue::glue("{teams %>% filter(team_abbr == input$posteam) %>% pull(team_logo_espn)}"),
        '">'
      )
    })
  
  output$picture2 <-
    renderText({
      c(
        '<img width="150" src="',
        glue::glue("{teams %>% filter(team_abbr == input$posteam) %>% pull(team_logo_espn)}"),
        '">'
      )
    })
  
  
  # say what the right decision is
  output$some_text <- renderText({ 
    
    return(glue::glue("<font size='+2'>Correct choice: <span style='color:red'>{tableData() %>% arrange(-choice_prob) %>% dplyr::slice(1) %>% pull(choice)}</span> (difference: 
      <span style='color:green'> <strong> {round(tableData() %>% arrange(-choice_prob) %>% dplyr::slice(1) %>% pull(choice_prob) - tableData() %>% arrange(-choice_prob) %>% dplyr::slice(2) %>% pull(choice_prob), 1)}%</strong></span>)</font>"))

  })
  
  output$some_text2 <- renderText({ 
    
    return(glue::glue("<font size='+2'>Correct choice: <span style='color:red'>{tableData2() %>% arrange(-choice_prob) %>% dplyr::slice(1) %>% pull(choice)}</span> (difference: 
      <span style='color:green'> <strong> {round(tableData2() %>% arrange(-choice_prob) %>% dplyr::slice(1) %>% pull(choice_prob) - tableData2() %>% arrange(-choice_prob) %>% dplyr::slice(2) %>% pull(choice_prob), 1)}%</strong></span>)</font>"))
    
  })
  
  
  observe({
    query <- parseQueryString(session$clientData$url_search)
    query1 <- paste(names(query), query, sep = "=", collapse=", ")
    # print(query1)
    # print(substr(query1, (nchar(query1) - 4), (nchar(query1))))
    if(substr(query1, (nchar(query1) - 4), (nchar(query1))) == "2pt=1"){
      updateTabsetPanel(session, "inTabset", selected = "two_pt")
    }
  })
  
  
}

# Create Shiny app ----
shinyApp(ui, server, enableBookmarking = "url")

